File: AndroidTest/AndroidTest/LoginBrowser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Android.App;
using Android.Content;
using Android.OS;
using Android.Runtime;
using Android.Views;
using Android.Widget;
using Android.Webkit;

namespace AndroidTest
{
	class LoginBrowser : WebViewClient
	{
		public delegate void GotAuthEventHandler (object sender, AuthVerifierArgs e);
		public event GotAuthEventHandler gotAuthVerifier;

		public override bool ShouldOverrideUrlLoading(WebView view, String url)
		{
			String callbackURL = "http://www.google.com";

			if (!url.Contains (callbackURL)) {
				Console.WriteLine (url);
				view.LoadUrl (url);
				return true;
			} else 
			{
				String authTokens = url.Substring (callbackURL.Length + 1);

				if (gotAuthVerifier != null) {
					AuthVerifierArgs authArgs = new AuthVerifierArgs ();

					int splitIndex = authTokens.IndexOf ("&");
					String authVerifier = authTokens.Substring (splitIndex);
					authVerifier = authVerifier.Substring (0, authVerifier.Length - 4);

					authArgs.AuthVerifier = authVerifier;
					gotAuthVerifier (this, authArgs);
				}

				return true;
			}
		}

	}

	//The event args to pass the authentication tokens
	public class AuthVerifierArgs : EventArgs
	{
		private String authVerifier;
		public String AuthVerifier
		{
			get
			{
				return authVerifier;
			}
			set
			{
				authVerifier = value;
			}
		}
	}
}

File: Tumble+/src/com/JIAT/Tumble/LoginBrowser.java
package com.JIAT.Tumble;

import android.util.Log;
import android.webkit.WebView;
import android.webkit.WebViewClient;

/**
 * Created with IntelliJ IDEA.
 * User: Arsen
 * Date: 8/1/13
 * Time: 12:15 PM
 * To change this template use File | Settings | File Templates.
 */

public class LoginBrowser extends WebViewClient{

    private String callbackUrl;

    CallbackListener listener;

    public LoginBrowser(CallbackListener listener)
    {
        callbackUrl = "http://www.google.com";

        this.listener = listener;
    }

    public boolean shouldOverrideUrlLoading(WebView view, String url)
    {
        //If the URL we're visiting isn't the callback url, just load it
         if(!url.contains(callbackUrl))
         {
            view.loadUrl(url);

            return true;

        }//Otherwise, we're going to get the verifier from the callback url
        else
        {
            String authTokens = url.substring(callbackUrl.length() + 1);

            int splitIndex = authTokens.indexOf("&");
            String authVerifier = authTokens.substring(splitIndex);
            authVerifier = authVerifier.substring(0, authVerifier.length() - 4);
            authVerifier = authVerifier.substring(16);

            listener.gotCallback(authVerifier);

            return true;
        }
    }

}
